### DISCUSSION TITLE
Multivariate Quantile Regression

### DISCUSSION THREAD

**Discussion by jobstdavid** [kogalur/randomForestSRC#273] [Q&A] [answered]
Hello, 
I got some short questions concerning the multivariate regression forests, especially concerning multivariate quantile regression forests in your package.

1. Concerning multivariate quantile regression: If I understood it right for multivariate quantile regression, you fit first a multivariate random forest. Afterwards quantiles can be calculated for each response separately. Are the CDF of each response and consequently the quantile function of each response determined as in the univariate quantile regression forest (by Meinhausen)? Why don’t you allow the splitting rule „mahalanobis“ for multivariate quantile regression forests? This splitting rule could also improve the quantile forecasts.
2. How is it possible to produce multivariate out of bag quantile forecasts with your R-package? For example I have the R-code

library(randomForestSRC)
fit <- quantreg(cbind(mpg, cyl) ~ ., data = mtcars, method = "forest")

Now, I would like to predict the quantiles = 1:50/51 for new predictor values X. How is this possible?
Thank you very much for your help!

David

**Comment by ishwaran** [comment, marked as answer]
Hi, thanks for your questions.

1. Yes, we first fit a multivariate random forest and then afterwards the quantiles are calculated separately for each response.  These quantiles are calculated as in the univariate setting.  Note that the method used for calculating quantiles can be varied using the option `method` and that the Meinhausen approach is not the default approach (as we and others have found better means to accomplish this).
2. It was an oversight to not allow `splitrule="mahalanobis"` and is easily corrected.  We will make that change in the next release to CRAN and thank you for pointing this out.  This will allow the user to specify more generally the multivariate splitting rule.
3. Regarding the prediction question.  On new test data prediction is not technically referred to as "out-of-bag".  Nevertheless, to answer your question, from the help file, to predict on new test data: 
    
     ```
## train (grow) call followed by test call
     o <- quantreg(mpg ~ ., mtcars[1:20,])
     o.tst <- quantreg(object = o, newdata = mtcars[-(1:20),])
     
     ## extract test set quantiles and conditional statistics
     print(get.quantile(o.tst))
     print(get.quantile.stat(o.tst))
     

```

**Reply by jobstdavid** [reply]
Thank you very much for your answer! 
Concerning 3: Of course I meant "out-of-sample" prediction and not "out-of-bag" prediction. I know, how to calculate quantiles for a univariate quantile regression forest. But how do I calculate the quantiles for each response in a multivariate quantile regression forest, e.g. for a bivariate quantile regression forest: 

`fit <- quantreg(cbind(mpg, cyl) ~ ., data = mtcars[1:20, ])
out <- predict(fit, mtcars[-c(1:20), ])
print(get.quantile(out, c(0.25)))`

The last line of the code 
`print(get.quantile(out, c(0.25)))`
does not yield the quantile for each response for alpha = 0.25. 

Can you maybe help me? Is there a build in function in this package?

And how can I choose the `method = "forest"` in rfsrc or predict for the weight calculation? Does it simply work with 
`method = "forest"`?

**Comment by ishwaran** [comment]
Keep in mind that all calculations have to be done within `quantreg`.  This is unlike a regular forest and therefore you should not be making calls to `predict` (i.e. `predict.rfsrc`) with an object created by `quantreg`.  So you have to follow very carefully the options for the function; and I recommend carefully reading the help file.

For your example, notice that we need to invoke options `object` and `newdata` for the prediction call:

```
fit <- quantreg(cbind(mpg, cyl) ~ ., data = mtcars[1:20, ]) 
out <- quantreg(object = fit, newdata = mtcars[-c(1:20), ]) 
```

Now when you make the call to `get.quantile` it will return a list populated by the quantiles, with one entry in the list for each of the outcomes (as stated in the help file: "_For multivariate forests, returned values will be a list of length equal to the number of target outcomes"_).

````
print(get.quantile(out, c(0.25)))
````

Regarding the call to `method`, this is done as follows:

````
fit <- quantreg(cbind(mpg, cyl) ~ ., data = mtcars[1:20, ], method = "forest")
````

Here's a link to the [quantreg help file](https://www.randomforestsrc.org/reference/quantreg.rfsrc.html)

**Reply by jobstdavid** [reply]
Thank you very much! And one last question: When I call `rfsrc` with `forest.wt = TRUE`, which method is used to calculate the forest weights?

**Comment by ishwaran** [comment]
There is only one method used to compute the forest weights.  This is because there is only one definition for the forest weights.

See the following vignette [for forest weights](https://www.randomforestsrc.org/articles/forestWgt.html) to see what the definition is.
